power_grid_model_c/power_grid_model/include/power_grid_model/optimizer/tap_position_optimizer.hpp

```diff
@@ -890,21 +911,18 @@ class TapPositionOptimizerImpl<std::tuple<TransformerTypes...>, StateCalculator,
             UpdateBuffer update_data;
             Idx rank_index = 0;
 
-            for (Idx i = 0; i < static_cast<Idx>(regulator_order.size()); ++i) {
-                auto const& same_rank_regulators = regulator_order[i];
-                for (Idx j = 0; j < static_cast<Idx>(same_rank_regulators.size()); ++j) {
-                    auto const& regulator = same_rank_regulators[j];
-                    BinarySearchOptions const options{strategy_max, Idx2D{i, j}};
-                    tap_changed =
-                        adjust_transformer(regulator, state, result, update_data, search, options) || tap_changed;
-                }
-                if (tap_changed) {
-                    iterations_per_rank[rank_index + 1] = 0;
-                    ++iterations_per_rank[rank_index];
-                    break;
-                }
-                ++rank_index;
-            }
+            auto const adjust_transformer_in_rank = [&](Idx const& rank_idx, Idx const& transformer_idx,
+                                                        std::vector<RegulatedTransformer> const& same_rank_regulators) {
+                auto const& regulator = same_rank_regulators[transformer_idx];
+                BinarySearchOptions const options{strategy_max, Idx2D{rank_idx, transformer_idx}};
+                tap_changed = adjust_transformer(regulator, state, result, update_data, search, options) || tap_changed;
+                return tap_changed;
+            };
+
+            RankIterator rank_iterator;
+            rank_iterator.iterate_ranks(regulator_order, adjust_transformer_in_rank, tap_changed, iterations_per_rank,
+                                        rank_index);
+
             if (tap_changed) {
                 if (static_cast<uint64_t>(iterations_per_rank[rank_index]) > 2 * max_tap_ranges_per_rank[rank_index]) {
```

Comment:
i think it's good to do the outer `while` loop + stop criterion in the `RankIterator` as well if possible

Reply:
Don't think I agree. The fact that while `tap_changed == true` there needs to happen something feels like behavior of `TapPositionOptimizerImpl`. The iterations itself can happen in the `RankIteration`.

Next to that I think moving the while loop + stop criteria would make thing overly complicated.

Reply:
ok